Dependencies/xr/Source/OpenXR/SceneUnderstanding.cpp

```diff
@@ -71,500 +73,463 @@ SceneUnderstanding::SceneUnderstanding::InitOptions::InitOptions(
 
-        throw std::exception(/*unknown mesh id*/);
-    }
+        args.RemovedSceneObjects.clear();
+        for (const auto& objectId : m_removedObjects)
+        {
+            args.RemovedSceneObjects.push_back(objectId);
+        }
+        m_removedObjects.clear();
 
-private:
-    const std::map<XrSceneObjectKindTypeMSFT, xr::SceneObjectType> s_sceneObjectTypeMap
-    {
-        {XrSceneObjectKindTypeMSFT::XR_SCENE_OBJECT_KIND_TYPE_BACKGROUND_MSFT, xr::SceneObjectType::Background},
-        {XrSceneObjectKindTypeMSFT::XR_SCENE_OBJECT_KIND_TYPE_CEILING_MSFT, xr::SceneObjectType::Ceiling},
-        {XrSceneObjectKindTypeMSFT::XR_SCENE_OBJECT_KIND_TYPE_FLOOR_MSFT, xr::SceneObjectType::Floor},
-        {XrSceneObjectKindTypeMSFT::XR_SCENE_OBJECT_KIND_TYPE_PLATFORM_MSFT, xr::SceneObjectType::Platform},
-        {XrSceneObjectKindTypeMSFT::XR_SCENE_OBJECT_KIND_TYPE_UNKNOWN_MSFT, xr::SceneObjectType::Unknown},
-        {XrSceneObjectKindTypeMSFT::XR_SCENE_OBJECT_KIND_TYPE_WALL_MSFT, xr::SceneObjectType::Wall}
-    };
+        args.UpdatedMeshes.clear();
+        for (const auto& meshId : m_updatedMeshes)
+        {
+            args.UpdatedMeshes.push_back(meshId);
+        }
+        m_updatedMeshes.clear();
 
-    enum class State
-    {
-        Ready,
-        ComputingScene,
-        Waiting
-    };
+        args.RemovedMeshes.clear();
+        for (const auto& meshId : m_removedMeshes)
+        {
+            args.RemovedMeshes.push_back(meshId);
+        }
+        m_removedMeshes.clear();
+
+        args.UpdatedPlanes.clear();
+        for (const auto& planeId : m_updatedPlanes)
+        {
+            args.UpdatedPlanes.push_back(planeId);
+        }
+        m_updatedPlanes.clear();
 
-    bool m_initialized{ false };
-    xr::SpaceHandle m_viewSpace{};
-    xr::DetectionBoundary m_detectionBoundary;
-    XrTime m_updateInterval;
-    xr::SceneObserverHandle m_sceneObserverHandle{};
-    XrTime m_lastSceneUpdate{};
-    State m_state{ State::Ready };
-    xr::SceneHandle m_sceneHandle{};
-    XrTime m_lastFrameUpdate{};
-    XrTime m_lastSceneObjectUpdate{};
-    std::vector<XrSceneObjectLocationMSFT> m_sceneObjectLocations;
+        args.RemovedPlanes.clear();
+        for (const auto& planeId : m_removedPlanes)
+        {
+            args.RemovedPlanes.push_back(planeId);
+        }
+        m_removedPlanes.clear();
+    }
 
+    std::unique_ptr<xr::su::SceneObserver> m_sceneObserver;
+    std::unique_ptr<xr::su::Scene> m_scene;
+    xr::SceneBounds m_sceneBounds{};
+    ScanState m_scanState{ ScanState::Idle };
+    XrTime m_updateInterval{ 0 };
+    XrTime m_lastUpdateTime{ 0 };
+
+    std::unordered_map<xr::su::SceneObject::Id, SceneObject::Identifier> m_objectIds{};
+    std::unordered_map<SceneObject::Identifier, std::unique_ptr<SceneObject>> m_objects{};
+    std::unordered_map<xr::su::SceneMesh::Id, Mesh::Identifier> m_meshIds{};
+    std::unordered_map<Mesh::Identifier, std::unique_ptr<Mesh>> m_meshes{};

```

Comment:
What is the purpose of having the values in these maps be wrapped in a `unique_ptr`?

Reply:
it lets me get access to the struct generated id before inserting into the maps. If you have a better suggestion, definitely let me know :)

Reply:
The auto generated id is only auto generated when you use the default constructor of `Mesh`. If you use the copy constructor or explicitly set all members then you could set the ID to whatever you want. I think you should be able to create a local `Mesh`, get its auto generated ID then insert or emplace the Mesh into the `unordered_map` using that ID. It should invoke the copy constructor of `Mesh` without incrementing the static `NEXT_ID`.

Change this:

```c++
auto mesh = std::make_unique<Mesh>();
const auto meshId = mesh->ID;
m_meshIds[xrMesh.id] = meshId;
m_meshes[meshId] = std::move(mesh);
```

to:

```c++
// Assuming m_meshes is updated to use Mesh instead of unique_ptr<Mesh>:
std::unordered_map<Mesh::Identifier, Mesh> m_meshes;

Mesh mesh; // auto generates new ID
const auto meshId = mesh.ID; // get the auto generated ID
m_meshes.emplace(meshId, mesh); // should call copy constructor of Mesh. NEXT_ID shouldn't change.
m_meshIds[xrMesh.id] = meshId;
```

Although I haven't tried this since I'm not setup to build this repo. 

You need to use insert or emplace instead of `operator[]` because of how `std::unordered_map::operator[]` behaves. `operator[]` will actually call the default constructor before then calling the copy assignment operator. The STL is full of weird stuff like this.

Reply:
this worked. thank you for the suggestion @dgough!